[addons][visualization] fix stuttering on high sample streams and fix missing data

AlwinEsch opened the pull request:
## Description

Before was on streams with very high samplerate e.g. 352800 together with high CPU consumption a very big continues stuttering on visualization show.

Commit 1:
----------------------
**[audioengine][viz] fix visualization frames amount**

It has used to big frame size from e.g. 384000 Hz to 44100 Hz
and the small Samplerate used the bigger frames amount from high rate.

Commit 2:
----------------------
**[audioengine][viz] fix visualization audio data size**

When copying the stream for visualization, its entire content was never used and only the number of samples was given and not multiplied by the required number of channel

Commit 3:
----------------------
**[audioengine] fix wrong PTS calc on resampler**

It has used by PTS correction, the sample amount of input where was related to e.g. 384000Hz sample rate but in resampler it has used the sample rate of output where e.g. 44100Hz to calculate and produce a much to high value.

## User related

- Fix: GUI visualization stuttering on high samplerate streams

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->

## Motivation and Context

This fix also this issue here https://github.com/xbmc/visualization.waveform/issues/17
Ans seen the bug also by new addon visualization.matrix.

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?

Was with visualization.wav and visualization.matrix on several highlevel streams and see videos below.

<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

Video with fix:
https://youtu.be/7QWBZ72TdZg

Video without fix:
https://youtu.be/fv4ZiYFs2hA


## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

fritsch commented on xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp:
```diff
@@ -2234,16 +2234,11 @@ bool CActiveAE::RunStages()
             while(!m_vizBuffers->m_outputSamples.empty())
             {
               CSampleBuffer *buf = m_vizBuffers->m_outputSamples.front();
-              if ((now - buf->timestamp) < 0)
```
Take care that the vizBuffers don't fill up. This is especially here to drop those that are already way, way behind? E.g. drop those that are > 0.5 seconds behind? Also keep in mind that user can see the delay in the waveform now. Else we could also just add random samples ;-)

FernetMenta commented:
congrats. you dropped the sync mechanism. you don't think that this code was added for any reason, do you?

AlwinEsch commented:
I look to improve complete about this (becomes bigger) but how now is unusable in visualization.

AlwinEsch commented:
Hi @FernetMenta 

Updated the request in a more corrective way.
And runs perfectly for tests.

However, in the resampling of the audio engine there really seems to be a problem with its PTS calculation.

Refers e.g. on this call to ffmpeg: https://github.com/xbmc/xbmc/blob/master/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAEResampleFFMPEG.cpp#L260-L261
There at **`av_rescale_rnd`** at the first value, this is done **`swr_get_delay(m_pContext, m_src_rate)`**. 

When I searched the web about this, I found: https://cpp.hotexamples.com/de/examples/-/-/av_rescale_rnd/cpp-av_rescale_rnd-function-examples.html

Almost everywhere, **`nb_samples`** is added to this in examples, had also tested with it in Kodi and partially corrected the problem as well and PTS was more correct, but then it threw some other things in the resampler upside down.

Unfortunately, this exceeds my knowledge and does not want to mess around with it, because then not only visualization would be affected.

AlwinEsch commented:
Hope the current way is OK, but definitely needs something to fix the problem, otherwise visualizations on thick streams cannot be used.

Strangely, with smaller e.g. 96kHz it is not immediately noticeable, only when you take a closer look, but with 352..8 e.g. it is most extreme.

AlwinEsch commented:
> congrats. you dropped the sync mechanism. you don't think that this code was added for any reason, do you?

@FernetMenta Sorry about my first attempt to fix, it was really mega stupid. But is this now OK or a idea about useful other solution?

Here is really buggy at the moment and not want to ignore the error further.

FernetMenta commented:
This change has more than just a single error. Apart from that, the refactoring ruins the design. Creating a function and calling it just a single time is bad coding style. Tearing the viz away and leaving the rest with an error (that's what you assume) is a very bad style too.

Issues should always be tracked down to the root cause. If you don't understand what's going wrong and why, you can't fix it. If you want to go that path, I can assist. Otherwise you are on your own.

AlwinEsch commented:
> Creating a function and calling it just a single time is bad coding style.

You mean the `bool ResamplerUsed() const { return m_resampler != nullptr; }` (I think yes)? Or that I separated on first commit the Visualization related parts to own functions?

> This change has more than just a single error.

Thats true, seen also here https://github.com/xbmc/xbmc/blob/master/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp#L2268 it goes that way.

Added also a note here https://github.com/xbmc/xbmc/pull/18294#issuecomment-676694915 about `nb_samples`, mostly everywhere where I found on other projects with ffmpeg resample use, that it was used to calculate the packet time.
Was it forgotten in our code or was there a specific reason to leave it out?

In the whole audioengine is by resample use the timestamp not correct, seems only to work that maybe somewhere (where I'm not found now) this becomes corrected or not relate to the audio output to speakers.

FernetMenta commented:
> In the whole audioengine is by resample use the timestamp not correct

I don't think so and I don't see any relationship to the example you linked.

AlwinEsch commented:
I will check again in other places, just find it strange if the timestamp starts nicely with 0 without resample. But with resample, which takes the timestamp of the following packet in the previous one.

AlwinEsch commented on xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAEBuffer.cpp:
```diff
@@ -296,18 +293,10 @@ bool CActiveAEBufferPoolResample::ResampleBuffers(int64_t timestamp)
-            m_lastSamplePts = in->timestamp;
-          else
-            in->pkt_start_offset = 0;
-        }
```
I removed the function value `int64_t timestamp`, as this was only used for visualizations and if unnecessary stream work is not used, it should be checked for" 0 ".

AlwinEsch reviewed (commented):
Hopefully I have now found an acceptable way, because the `pkt_start_offset` was really wrong before.

FernetMenta commented on xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp:
```diff
@@ -2222,23 +2224,31 @@ bool CActiveAE::RunStages()
                 memcpy(viz->pkt->data[i], out->pkt->data[i], bytes);
               }
               viz->pkt->nb_samples = samples;
+              viz->pkt_start_offset = out->pkt_start_offset;
```
> Think to found the main bug about.
wrong. this change doesn't have any impact on the current code because timestamps of audiostream are not used for viz. pkt_start_offset is set to 0 in ResampleBuffers

AlwinEsch replied:
The `timestamps` of given audio streams are always 0. 
Here it do mostly the same https://github.com/xbmc/xbmc/blob/master/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp#L1960 to get a PTS about (but never called as 0 on `buf->timestamp`.

But it cannot understand why this is not needed, in the resample it should correspond to the data from the input where this `pkt_start_offset` is missing from the copy.

The error is then even without a resample, only this value is so small that it is not noticeable.

AlwinEsch replied:
Here about the current problem, with values prcessed inside:
```cpp
      int out_samples = m_resampler->Resample(m_planes,
                                              m_procSample->pkt->max_nb_samples - m_procSample->pkt->nb_samples,
                                              in ? in->pkt->data : NULL,
                                              in ? in->pkt->nb_samples : 0,
                                              m_resampleRatio);
      //----------> 1. out_samples = 2173
      //----------> 2. out_samples = 2205
      //----------> 3. out_samples = 2205
      // ... Stays the same
       
      // in case of error, trigger re-create of resampler
      if (out_samples < 0)
      {
        out_samples = 0;
        m_changeResampler = true;
      }
      
      m_procSample->pkt->nb_samples += out_samples;
      //----------> 1. m_procSample->pkt->nb_samples: 2173 = 0 + 2173
      //----------> 2. m_procSample->pkt->nb_samples: 2205 = 0 + 2205
      //----------> 3. m_procSample->pkt->nb_samples: 2205 = 0 + 2205
      // ... Stays the same

      busy = true;
      m_empty = (out_samples == 0);

      if (in)
      {
        if (!timestamp)
        {
          if (in->timestamp)
            m_lastSamplePts = in->timestamp;
          else
            in->pkt_start_offset = 0;
        }
        else
        {
          m_lastSamplePts = timestamp;
          // Here how CPU clock brings wrong values about "timestamp" and destroy use
          //----------> 1. 16610 (correct 50 jump)
          //----------> 2. 16710 (goes 100 one step time overjumped, comes much times)
          //----------> 3. 16760 (correct 50 jump)
          //----------> ... ~50 on step (correct)
          //----------> 7. 16961 (correct 50 jump)
          //----------> 8. 17412 (goes 451 forward, also much times)
          //----------> 9. 17512 (goes 100 one step time overjumped, comes much times)
          //----------> 10. 17561 (correct 50 jump)
          // This time jumps always continued inside.
          
          in->pkt_start_offset = 0;
        }

        // pts of last sample we added to the buffer
        m_lastSamplePts += (in->pkt->nb_samples-in->pkt_start_offset) * 1000 / m_format.m_sampleRate;
        // Here where "pkt_start_offset" is always "0" (set short before) and the timestamp from input becomes always 400 added (where destroy the show time)
        // as a Paket have 50 in his size the audio engine must process additional 8 until it reachs the first for give to addon
        //----------> 1.  m_lastSamplePts: 17010 = 16610 + (17640 - 0) * 1000 / 44100
        //----------> 2.  m_lastSamplePts: 17110 = 16710 + (17640 - 0) * 1000 / 44100 <!-----
        //----------> 3.  m_lastSamplePts: 17160 = 16760 + (17640 - 0) * 1000 / 44100
        //----------> 4.  m_lastSamplePts: 17210 = 16810 + (17640 - 0) * 1000 / 44100
        //----------> 5.  m_lastSamplePts: 17260 = 16860 + (17640 - 0) * 1000 / 44100
        //----------> 6.  m_lastSamplePts: 17310 = 16910 + (17640 - 0) * 1000 / 44100
        //----------> 7.  m_lastSamplePts: 17361 = 16961 + (17640 - 0) * 1000 / 44100
        //----------> 8.  m_lastSamplePts: 17812 = 17412 + (17640 - 0) * 1000 / 44100 <!-----
        //----------> 9.  m_lastSamplePts: 17912 = 17512 + (17640 - 0) * 1000 / 44100 <!-----
        //----------> 10. m_lastSamplePts: 17961 = 17561 + (17640 - 0) * 1000 / 44100
      }

      // calculate pts for last sample in m_procSample
      int bufferedSamples = m_resampler->GetBufferedSamples();
      //----------> 1. bufferedSamples = 32
      //----------> 2. bufferedSamples = 32
      //----------> 3. bufferedSamples = 32
      // ... Stays the same
      
      m_procSample->pkt_start_offset = m_procSample->pkt->nb_samples;
      //----------> 1. m_procSample->pkt_start_offset = 2173 (m_procSample->pkt->nb_samples)
      //----------> 2. m_procSample->pkt_start_offset = 2205 (m_procSample->pkt->nb_samples)
      //----------> 3. m_procSample->pkt_start_offset = 2205 (m_procSample->pkt->nb_samples)
      // ... Stays the same
      
      m_procSample->timestamp = m_lastSamplePts - bufferedSamples * 1000 / m_format.m_sampleRate;
      //----------> 1. m_procSample->timestamp: 17010 = 17010 - 32 * 1000 / 44100
      //----------> 2. m_procSample->timestamp: 17110 = 17110 - 32 * 1000 / 44100 <!-----
      //----------> 3. m_procSample->timestamp: 17160 = 17160 - 32 * 1000 / 44100
      //----------> 4. m_procSample->timestamp: 17210 = 17210 - 32 * 1000 / 44100
      //----------> 5. m_procSample->timestamp: 17260 = 17260 - 32 * 1000 / 44100
      //----------> 6. m_procSample->timestamp: 17310 = 17310 - 32 * 1000 / 44100
      //----------> 7. m_procSample->timestamp: 17361 = 17361 - 32 * 1000 / 44100
      //----------> 8. m_procSample->timestamp: 17812 = 17812 - 32 * 1000 / 44100 <!-----
      //----------> 9. m_procSample->timestamp: 17912 = 17912 - 32 * 1000 / 44100 <!-----
      //----------> 10. m_procSample->timestamp: 17961 = 17961 - 32 * 1000 / 44100
      
```
Is from "CActiveAEBufferPoolResample::ResampleBuffers"

Where should be the place to correct the the wrongly in future shown timestamp without use of `pkt_start_offset`?

AlwinEsch replied:
What also works without set of `pkt_start_offset`in AudioEngine, to use for `m_lastSamplePts` the `m_procSample->pkt` instead of `in->pkt`
```cpp
        // pts of last sample we added to the buffer
        int64_t samples = m_procSample->pkt->nb_samples-m_procSample->pkt_start_offset;
        if (samples > 0)
          m_lastSamplePts += samples * 1000 / m_format.m_sampleRate;
```

Is that more correct as the note `// pts of last sample we added to the buffer` also rever to avbout add to buffer where is `m_procSample`?

FernetMenta replied:
Before blaming code you obviously do not fully understand and trying to change things in a try and error fashion, you should really try to answer this questions: Why does it work with normal sample rates and why do you observe issues with high sample rates. 

```
         m_lastSamplePts = timestamp;
          // Here how CPU clock brings wrong values about "timestamp" and destroy use
          //----------> 1. 16610 (correct 50 jump)
          //----------> 2. 16710 (goes 100 and skipped a step, comes often)
```

Jumps in timestamp here are either a result of GetDelay of audio sink or AEDelayStatus::maxcorrection prevents correction of delay with current time (maxcorrection might be zero because it is not set)

AlwinEsch commented:
How should the error be corrected now?

I know where caused by, can prove it by means of a video (on YouTube) and a value list (see comment https://github.com/xbmc/xbmc/pull/18294#discussion_r476692649).

Knows what would fix it.

But can't mind reading how you would have liked to fix it.

Could you please give a sample code where and how the fix should be for you?

I don't want to ignore this error and pull request here and fix it as soon as possible.

AlwinEsch committed:
[audioengine][viz] fix visualization frames amount where

It has used to big frame size from e.g. 384000 Hz to 44100 Hz
and the small Samplerate used the bigger frames amount from high rate.

AlwinEsch committed:
[audioengine][viz] fix visualization audio data size

When copying the stream for visualization, its entire content was never used and only the number of samples was given and not multiplied by the required number of channel

AlwinEsch commented:
It's updated again. Regarding time,

it used too large a number of samples in the buffer during resample, which means that its time was defined as 400ms per packet instead of 50ms (everywhere else in Audioengine). This always reached the end too quickly and created the stutter.

The number of frames which belong to 384000 then to use at 44100 does not quite fit.

---------------------------------

However, because of the delay, there must be something at `pkt_start_offset` otherwise it will always be in the future.

Without resample, this error does not occur, because `timestamp` is not changed there (only copied).

Assume this should be reduced again with `int bufferedSamples = m_resampler-> GetBufferedSamples ();` and used correctly, but since this only returns 32, this only comes out as a "0" calculation in the integer and PTS remains in the future.

Please explain how to fix this correctly!
I also want to learn, but not always dismissed it with complaints.

AlwinEsch commented:
Everything should now be correct on the request. In the resampler it had also calculated the amount of samples from the input (384000Hz) with the sample rate from the output (44100Hz), which of course is much too high.

This now also performs the correct calculation there.

Would have been easier to recognize if `AEAudioFormat m_format;` was for example called `AEAudioFormat m_outputFormat;`.

@FernetMenta @fritsch is it correct now?

FernetMenta commented on xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAEBuffer.cpp:
```diff
@@ -310,7 +310,8 @@ bool CActiveAEBufferPoolResample::ResampleBuffers(int64_t timestamp)
         // pts of last sample we added to the buffer
-        m_lastSamplePts += (in->pkt->nb_samples-in->pkt_start_offset) * 1000 / m_format.m_sampleRate;
+        m_lastSamplePts +=
+            (in->pkt->nb_samples - in->pkt_start_offset) * 1000 / m_inputFormat.m_sampleRate;
```
nice finding. maybe use in->pkt->config.sample_rate instead of m_inputFormat.m_sampleRate. that would make it clear that all values belong to the in-buffer

AlwinEsch replied:
Has changed it to use `in->pkt->config.sample_rate` and match more to the other values there, thanks.

FernetMenta commented on xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp:
```diff
@@ -2238,7 +2241,8 @@ bool CActiveAE::RunStages()
               {
-                unsigned int samples = static_cast<unsigned int>(buf->pkt->nb_samples);
+                unsigned int samples = static_cast<unsigned int>(buf->pkt->nb_samples) *
+                                       buf->pkt->config.channels / buf->pkt->planes;
```
correct. this change was missed when planar formats were added.

FernetMenta commented:
looks good to me now. good work!

AlwinEsch committed:
[audioengine] fix wrong PTS calc on resampler

It has used by PTS correction, the sample amount of input where was related
to e.g. 384000Hz sample rate but in resampler it has used the sample rate of
output where e.g. 44100Hz to calculate and produce a much to high value.
